src/systemcmds/bsondump/bsondump.cpp

```diff
@@ -82,58 +91,224 @@ static int bson_print_callback(bson_decoder_t decoder, bson_node_t node)
 	return -1;
 }
 
+static int
+bson_docsize_callback(bson_decoder_t decoder, bson_node_t node)
+{
+
+	if (node->type == BSON_EOO) {
+		PX4_DEBUG("end of parameters");
+		return 0;
+	}
+
+	return 1;
+}
+
+int copy_file(const char *src_path, const char *dst_path)
+{
+	int src_fd = open(src_path, O_RDONLY);
+	int dst_fd = open(dst_path, O_WRONLY | O_TRUNC);
+
+	if (src_fd == -1 || dst_fd == -1) {
+		perror("Failed to open files for copying");
+
+		if (src_fd != -1) { close(src_fd); }
+
+		if (dst_fd != -1) { close(dst_fd); }
+
+		return -1;
+	}
+
+	char buffer[1024];
+	ssize_t bytes_read;
+
+	while ((bytes_read = read(src_fd, buffer, sizeof(buffer))) > 0) {
+		if (write(dst_fd, buffer, bytes_read) != bytes_read) {
+			perror("Failed to copy file");
+			close(src_fd);
+			close(dst_fd);
+			return -1;
+		}
+	}
+
+	close(src_fd);
+	close(dst_fd);
+
+	if (bytes_read == -1) {
+		perror("Error reading from source file");
+		return -1;
+	}
+
+	return 0;
+}
+
 extern "C" __EXPORT int bsondump_main(int argc, char *argv[])
 {
-	if (argc != 2) {
-		print_usage();
+	if (argc < 2) {
+		print_usage("Invalid number of arguments.");
+		return -1;
+	}
+
+	char *command = argv[1];
+	char *file_name = argv[3];
```

Comment:
This requires `argc >= 4`, but is overwritten further down, so you can move it inside.

Reply:
we have two valid commands, bsondump file_name and bsondump docsize file_name

Isn't that argc > 2? 2 argc for the first one and 3 for the second? What I'm I missing?

Reply:
Do you mean to move char *file_name to inside each case?

Reply:
fixed